apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/TransactionPool.cs

```diff
@@ -0,0 +1,247 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Google.Cloud.Spanner.V1.Logging;
+using Google.Protobuf;
+
+namespace Google.Cloud.Spanner.V1
+{
+    /// <summary>
+    /// The TransactionPool works in conjuntion with the Session pool to warm pooled Sessions
+    /// with the transaction semantics it last used (if appropriate).
+    /// When requesting a pooled session, the request can include transactionoptions which
+    /// is then used as a key in the sessionpool.
+    /// </summary>
+    public static class TransactionPool
+    {
+        // Holds transaction state on sessions, including the last created transaction and options used.
+        static readonly ConcurrentDictionary<Session, SessionInfo> s_sessionInfoTable = new ConcurrentDictionary<Session, SessionInfo>();
+
+        // Holds a list of all active transactions with back pointers to the owning session.
+        static readonly ConcurrentDictionary<ByteString, Session> s_activeTransactionTable = new ConcurrentDictionary<ByteString, Session>();
+
+        /// <summary>
+        /// Creates a transaction with the given transactoin options.
+        /// Depending on the state of the session, this may or may not result in a request
+        /// to Spanner over the network.
+        /// </summary>
+        /// <param name="session"></param>
+        /// <param name="client"></param>
+        /// <param name="options"></param>
+        /// <returns></returns>
+        public static async Task<Transaction> BeginPooledTransactionAsync(this SpannerClient client, Session session,
+            TransactionOptions options)
+        {
+            var info = s_sessionInfoTable.GetOrAdd(session, s => new SessionInfo {SpannerClient = client});
+        
+            //we need to await for previous task completion anyway -- otherwise there is a bad race condition.
+            await info.WaitForPreWarmAsync().ConfigureAwait(false);
+            
+            //now we see if we can return the prewarmed transaction.
+            // a lot has to be correct for us to return the precreated transaction.
+            // we have to have a valid transaction object.
+            // it needs to exist in the activetransactiontable to indicate that it hasn't been closed
+            // the options on it must equal the passed in options.
+            if (info.ActiveTransaction != null
```

Comment:
No -- this is much simpler.  BeginTransaction will only be called from the single owner of the session and should not be called concurrently on multiple threads.  (kind of like -- why would you call the REST api concurrently on the same session).
The reason this code exists is to support some optimizations the java client added -- which is to "prewarm" the session with some transaction semantics (we use the same transaction options as was used last time on the pooled session).  This saves the next transaction from waiting a full server roundtrip, but requires us to cache and check the prewarmed state to ensure it matches.
Note that if the transaction options currently set on the session do not match for whatever reason, they can simply be blown over by a subsequent call to being a transaction.